### ISSUE TITLE
Self Reference foreign keys are not generated

### ISSUE THREAD

**Issue by kderme** [yesodweb/persistent#1169]
A migration like this:
```
Chain
    previousChain ChainId Maybe
    deriving Show Eq
```
will fail to generate any foreign keys at postgreSQL. I have narrowed down the bug to the `getAddReference` function, at the guard `guard $ table /= s && cname /= fieldDB (entityId entity)`, especially the first check. Removing that fixes the problem. I think the check was added for the sql=id option https://github.com/yesodweb/persistent/pull/1066, so simply removing it may break other stuff?

**Comment by parsonsmatt** [comment]
Good catch thanks!

The specific line of code you're referencing is [this](https://github.com/yesodweb/persistent/pull/1066/files#diff-3a6af517565f4d723a74260939388b56ca107830b6a9ae68db5a877ca084f520R1122), I think. The reason I did that is because I wanted to forbid creating a foreign key from a table's `id` column to itself - something like

```
ALTER TABLE foo 
  ADD CONSTRAINT id REFERENCES foo(id)
```

I assumed that `cname :: DBName` was the name of the column that we were possibly creating a reference *from*. This bug report indicates that it's actually the column name of the *target* column. Which is a bit surprising to me!

**Comment by kderme** [comment]
Yes this is the line I meant. There is another line https://github.com/yesodweb/persistent/pull/1066/files#diff-3a6af517565f4d723a74260939388b56ca107830b6a9ae68db5a877ca084f520R1064 which also causes the problem. It's a similar case, but this triggers when there is a migrations on an existing table, while the other when the table is new.

**Comment by hdgarrood** [comment]
> The specific line of code you're referencing is [this](https://github.com/yesodweb/persistent/pull/1066/files#diff-3a6af517565f4d723a74260939388b56ca107830b6a9ae68db5a877ca084f520R1122), I think. The reason I did that is because I wanted to forbid creating a foreign key from a table's id column to itself - something like

> ```
> ALTER TABLE foo 
>  ADD CONSTRAINT id REFERENCES foo(id)
> ```

Do you remember what kind of Persistent entity definition might cause that? I guess something like this?

```
SelfFKPrimary
    Id SelfFKPrimaryId
```

In my local tests, this results in:

```haskell
    EntityDef
        { entityHaskell = EntityNameHS{unEntityNameHS = "SelfFKPrimary"}
        , entityDB = EntityNameDB{unEntityNameDB = "self_f_k_primary"}
        , entityId =
            EntityIdField
                ( FieldDef
                    { fieldHaskell = FieldNameHS{unFieldNameHS = "Id"}
                    , fieldDB = FieldNameDB{unFieldNameDB = "id"}
                    , fieldType = FTTypeCon Nothing "SelfFKPrimaryId"
                    , fieldSqlType = SqlInt64
                    , fieldAttrs = []
                    , fieldStrict = True
                    , fieldReference = NoReference
                    , fieldCascade = FieldCascade{fcOnUpdate = Nothing, fcOnDelete = Nothing}
                    , fieldComments = Nothing
                    , fieldGenerated = Nothing
                    , fieldIsImplicitIdColumn = True
                    }
                )
        , entityAttrs = []
        , entityFields = []
        , entityUniques = []
        , entityForeigns = []
        , entityDerives = []
        , entityExtra = fromList []
        , entitySum = False
        , entityComments = Nothing
        , entitySpan =
            Just [...]
        }
```

Notice in particular that `entityForeigns` is empty and `fieldReference` is set to `NoReference`, so I don't think persistent would ever attempt to create that FK constraint anyway?

**Comment by hdgarrood** [comment]
I think that's already handled inside `mkUnboundEntityDef`:

```haskell
        , unboundPrimarySpec =
            case (idField, primaryComposite) of
                (Just{}, Just{}) ->
                    error "Specified both an ID field and a Primary field"
                (Just a, Nothing) ->
                    if unboundIdType a == Just (mkKeyConType (unboundIdEntityName a))
                        then
                            DefaultKey (FieldNameDB $ psIdName ps)
                        else
                            SurrogateKey a
                (Nothing, Just a) ->
                    NaturalKey a
                (Nothing, Nothing) ->
                    DefaultKey (FieldNameDB $ psIdName ps)
```
specifically in the true branch of `if unboundIdType a == Just (mkKeyConType (unboundIdEntityName a))`. In that case we'll get a `DefaultKey`, which then delegates to `mkAutoIdField`, which always sets `fieldReference = NoReference`, so I don't think there's any situation in which this guard is necessary
